applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp

```diff
@@ -0,0 +1,655 @@
+     * @brief Default Constructor
+     * @details Initiliazes the flags
+     */
+    /// Constructor.
+    explicit ModalDerivativeScheme(Parameters InputParameters)
+    : 
+    Scheme<TSparseSpace,TDenseSpace>()
+    {
+        KRATOS_TRY
+
+        std::string finite_difference_type = InputParameters["finite_difference_type"].GetString();
+        if (finite_difference_type == "central")
+            mFiniteDifferenceTypeFlag = true;
+        else if (finite_difference_type == "forward")
+            mFiniteDifferenceTypeFlag = false;
+        else
+            KRATOS_ERROR << "\"finite_difference_type\" can only be \"forward\" or \"central\""  << std::endl;
+
+        mFiniteDifferenceStepSize = InputParameters["finite_difference_step_size"].GetDouble();
+
+        KRATOS_CATCH("")
+    }
+
+    /// Destructor.
+    ~ModalDerivativeScheme() override 
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief unction to be called when it is needed to initialize an iteration. It is designed to be called at the beginning of each non linear iteration
+     * @note Take care: the elemental function with the same name is NOT called here.
+     * @warning Must be defined in derived classes
+     * @details The function is called in the builder for memory efficiency
+     * @param rModelPart The model part of the problem to solve
+     * @param A LHS matrix
+     * @param Dx Incremental update of primary variables
+     * @param b RHS Vector
+     */
+    void InitializeNonLinIteration(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY
+
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        // Definition of the first element iterator
+        const auto it_elem_begin = rModelPart.ElementsBegin();
+
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(rModelPart.Elements().size()); ++i) {
+            auto it_elem = it_elem_begin + i;
+            it_elem->InitializeNonLinearIteration(r_current_process_info);
+        }
+
+        // Definition of the first condition iterator
+        const auto it_cond_begin = rModelPart.ConditionsBegin();
+
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(rModelPart.Conditions().size()); ++i) {
+            auto it_cond = it_cond_begin + i;
+            it_cond->InitializeNonLinearIteration(r_current_process_info);
+        }
+
+        // Definition of the first constraint iterator
+        const auto it_const_begin = rModelPart.MasterSlaveConstraintsBegin();
+
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(rModelPart.MasterSlaveConstraints().size()); ++i) {
+            auto it_const = it_const_begin + i;
+            it_const->InitializeNonLinearIteration(r_current_process_info);
+        }
```

Comment:
```suggestion
                EntitiesUtilities::InitializeNonLinearIterationAllEntities(rModelPart);
```

see https://github.com/KratosMultiphysics/Kratos/commit/03c675d053fbd0ade0f799d8c3575186730ea63e